src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageParser.g4

```diff
@@ -688,6 +688,7 @@ expression
 
 expr
     : primary                                                              #primaryExp
+    | expr DOT templateArgument                                            #templateExp
```

Comment:
Others below this have `bop=DOT`. Any reason this doesn't?

Also, I realize this works and mostly just curious, but how do you know this goes on the top versus the middle/bottom?

Reply:
>Others below this have bop=DOT. Any reason this doesn't?

Alternatives with `bop=X` are all binary operators, that are handled by a similar method; we name the operator so that we can reference `bop` instead of having to know what the operator is (we don't care). In this case, we have no need to have such a reference. Plus, the `.` isn't a binary operator in this context.

>Also, I realize this works and mostly just curious, but how do you know this goes on the top versus the middle/bottom?

Left recursive rules are a bit tricky, but this MOSTLY follows operator precedence except in a few rare cases. I was not able to fully refactor our `primary` rule when I did ANTLR2 -> ANTLR4 due to some existing weird stuff. Because of this, the grammar would get a bit messier if I put this alternative where it belongs (in `primary`). So, I did the next best thing; made this alternative the first one after `primary` in the `expr` rule instead of the last alternative in the `primary` rule.

Reply:
Thanks, it is clear now. I don't see how we will find out if the recursive rules matching fails except from user reporting.

Reply:
I am confident in this update, chances of a false match are basically nil since we have introduced new tokens that are radically different than anything else.